source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length)
+      : length_(length), slice_(new Envoy::Buffer::OwnedImpl()) {
```

Comment:
I changed the class definition a bit:
Since mem slice doesn't need to share ownership of slice_ with other instance, slice_ doesn't need to be a shared_ptr. 
Combined allocation of fragment and buffer itself.
Now there are only 2 allocations in this constructor.